### ISSUE TITLE
HOST all_to_all_v: a rank's own staged input appears in its receive window with no transfer (self slot)

### ISSUE THREAD

**Issue by georgebisbas** [hw-native-sys/pypto#2546]
## Summary

On the HOST rail of `pld.tensor.all_to_all_v`, for the **self destination** (`dest == my_rank`), row *k* of the receiver's window is observed holding that rank's own staged `input_buf` row *k* — **even when the collective transfers nothing at all**.

Found on Ascend 910B (`a2a3`, P=2) while adding skew coverage in #2536. Does **not** reproduce in `a2a3sim`.

## Why it cannot be an over-send

For `send_counts[dest] <= 0` the clamped count is `0`, and `recv_counts` correctly reads back `0`. The generated exchange kernel then computes:

```cpp
int64_t block_numel = rows64 * row_numel;          // 0 * 64 = 0
for (int64_t base = 0; base < block_numel; base += tile_cols) {
    ...                                            // body never executes
}
```

No TPUT is issued for that destination. Yet the data is present in the receive window.

## What has been ruled out

**Window aliasing.** A probe with no collective involved — write a sentinel into window A, read window B which is never written — reports the two `alloc_window_buffer` allocations are **distinct** on both sim and NPU (0/128 sentinel hits). `input_buf` and `data_buf` do not share storage.

```python
buf_a = pld.alloc_window_buffer(SIZE * pl.FP32.get_byte())
buf_b = pld.alloc_window_buffer(SIZE * pl.FP32.get_byte())
# fill A with a sentinel on every rank; read B on every rank; B never written
# -> B contains no sentinel bytes, on a2a3sim and on a2a3
```

So the mechanism is **unexplained**.

## Impact

No correctness impact is demonstrated. `recv_counts` bounds what any correct consumer reads, and every payload-row and count assertion passes on both rails, on sim and NPU. This is filed because data appearing in a receive window without a corresponding transfer is worth understanding, not because anything is known to be broken.

## Where it surfaces

`tests/st/distributed/test_l3_host_tensor_all_to_all_v.py::TestL3HostTensorAllToAllVSkew::test_host_all_to_all_v_skewed_counts`, cases `[2-negative]`, `[2-one_and_full]`, `[2-zero_and_full]`, all at `rank=0 src=0`. `[2-over_capacity]` passes because it has no padding rows.

That test's surplus-row assertion now **skips `src == rank`**, because on the self slot "the sender's surplus arrived" and "this rank's own staged data was already there" are the same bytes and the check cannot separate them. The exclusion references this issue.

The **InCore rail keeps the assertion and passes it**, self slot included, on NPU — it takes its input as a plain `pl.Tensor` rather than through a staged window, which is consistent with this being a property of the HOST staging path.

## Suggested next step

Dump the resolved addresses of `input_buf`, `data_buf` and the `CommRemotePtr(..., dest == my_rank)` target on NPU and compare, to find what writes the receive window outside the TPUT path.

**Comment by YunjiQin** [comment]
Closing this: the observation has a measurement explanation, and the receive window is clean.

## The bytes were never read from the window

`consume_step` copies only the valid rows:

```python
for r in pl.range(n_rows):          # n_rows = recv_counts[src]
    out = pl.store(pl.load(data, [base + r, 0], [1, SIZE]), [base + r, 0], out)
```

but the surplus-row assertion inspects `outputs[rank, base + k]` for `k >= n_rows` — rows the kernel never writes.

`pl.Out` lowers to `TensorArgType.OUTPUT_EXISTING` (`DistributedCodegen::ParamDirectionToTensorArgType`), and `task_args.h` grants it WRITE only:

```
//   INPUT -> READ, OUTPUT_EXISTING -> WRITE, INOUT -> READWRITE
```

Write-only means the host buffer is never uploaded, so `torch.zeros` never reaches the device; the whole device buffer is copied back afterwards. Those rows were therefore undefined device memory — not `data_buf`. There is no data path from the window's tail to `outputs` in that version of the test.

That also explains why it looked self-slot-specific rather than random. If the recycled `out` buffer held this rank's own staged input (same shape, same `[total, SIZE]` FP32 size), the observed value is `inputs[rank, src*mr+k]` while the assertion compares against `inputs[src, rank*mr+k]`. Under the payload formula `salt + r*1000 + d*100 + k*10 + j%10` those are equal exactly when `rank == src`.

The window-overlap probe is sound but tests the wrong pair: it compares two `alloc_window_buffer` allocations, and the suspect memory is the `pl.Out` device buffer, not another window.

## Verified after making the window observable

georgebisbas/pypto@b3383dcd (pushed to #2536) copies the tail rows into `out` alongside the valid ones on both rails, so `outputs` mirrors the whole window, and turns the tail check into a positive assertion — the runtime zeroes a comm-domain window at allocation before publishing the handle to peers (`aclrtMemset` in `comm_hccl.cpp`'s `alloc_domain`), so a row no TPUT ever wrote must read exactly `0`:

```python
assert torch.all(got_row == 0.0)
```

The `src == rank` exclusion is removed — the self slot is checked like every other pair.

Result on Ascend 910B via task-submit, both rails, P=2 and P=4, including `[2-negative]`, `[2-one_and_full]` and `[2-zero_and_full]` at `rank=0 src=0`: **20/20**, and the same on `a2a3sim`.

Note the assertion is exactly the negation of this issue's claim on the self slot. `would_have_been_sent` there is `inputs[rank, rank*mr+k]`, which is byte-identical to the staged `input_buf` row this issue reports seeing. Had that data been in the window, the check would fail. It reads `0`.

## What this does and does not settle

Settled: on the HOST rail, self destination, `send_counts <= 0`, the receive window rows are zero on NPU. No write reaches the window outside the TPUT path, so there is nothing to chase in the "suggested next step" direction.

Not settled by this alone: sample size is one full-suite run per platform. If the zero assertion ever flakes in CI, that would be the first real evidence for a stray write — and it would now point at the window rather than at an unwritten output buffer.
